File: src/test/resources/RemoveDups.java
import java.util.*;
import org.algohub.engine.type.*;

public class Solution {
    /**
     * @param linkedList Linked List where we need to remove duplicates
     * @return  Linked List with removed duplicates
     */
    public LinkedListNode<Integer> removeDuplicates(LinkedListNode<Integer> node) {
        LinkedListNode<Integer> root = node;
        HashSet<Integer> set = new HashSet<>();
        LinkedListNode<Integer> previous = null;

        while (node != null) {
            if (set.contains(node.value)) {
                previous.next = node.next;
            } else {
                set.add(node.value);
                previous = node;
            }
            node = node.next;
        }

        return root;
    }
}

File: src/test/resources/RemoveDups.rb
require 'set'

class Solution
  def removeDuplicates(node)
    root = node
    set = Set.new

    previous = nil

    while node
      if set.include?(node.value)
        previous.next = node.next
      else
        set << node.value
        previous = node
      end

      node = node.next
    end

    root
  end
end
